torch/csrc/utils/future.h

```diff
@@ -133,6 +133,19 @@ class TORCH_API Future final {
     addCallback([this, cb]() { cb(*this); });
   }
 
+  template<typename R>
+  std::shared_ptr<Future<R>> then(std::function<R(const Future<T>& future)> cb) {
```

Comment:
I am not sure if this is the appropriate API. Since the returned Future represents the completion of the callback, it should also hold the returned value of the callback?

Reply:
What's the reason for the CB taking in a future instead of the completed value of the future? Is it so that we can access errors on the future when running the callback?

Reply:
> what's the reason for the CB taking in a future instead of the completed value of the future? Is it so that we can access errors on the future when running the callback (if so see above comment)?

Yep, the return value `T` might not be available if the callback crashed.